### ISSUE TITLE
torch.nn

### ISSUE THREAD

**Issue by Ludobico** [Ludobico/Python_AL_01#31]
# torch.nn
`torch.nn`은 Pytorch의 신경망(Neural Netrowk) 모듈입니다. 이 모듈은 신경망을 정의하고 학습하는데 필요한 여러 기능을 제공합니다.

`torch.nn`은 신경망을 정의할 때 `nn.Module` 클래스를 상속하는데, 이 클래스는 신경망의 구조를 정의할 수 있게 해줍니다. 신경망의 구조는 일반적으로 층(Layer)로 구성되며, 층은 입력값을 처리하고 출력값을 생성합니다. torch.nn은 신경망의 각 층을 nn.Module의 서브클래스로 정의할 수 있게 해줍니다.

torch.nn은 신경망을 학습할 때도 유용합니다. 신경망을 학습할 때는 `손실 함수(Loss function)`과 `최적화 알고리즘(Optimization algorithm)`이 필요합니다.

### 손실 함수
손실 함수(Loss function)은 신경망의 출력과 정답의 차이를 측정하는 함수입니다. 손실 함수의 값은 신경망의 성능을 나타냅니다. 손실 함수의 값을 최소하하는 것은 신경망의 성능을 최대화 하는 것과 같습니다.

### 최적화 알고리즘
최적화 알고리즘은 신경망의 가중치와 편향을 조정하여 손실 함수의 값을 최소화하는 것을 목표로 합니다. 신경망을 학습할 때는 손실 함수의 값을 최소화하기 위해 최적화 알고리즘을 사용합니다.

최적화 알고리즘을 사용할 때는 최적화 알고리즘 객체를 생성한 다음, 최적화할 모델의 가중치를 지정해주고, 손실 함수의 값을 측정한 후 최적화 알고리즘 객체의 `step` 메서드를 호출해주면 됩니다.

# torch.nn.Linear
`torch.nn.Lienar`는 Pytorch의 신경망 모듈에 정의된 선형 계층(Linear layer)입니다. 선형 계층은 입력값과 가중치를 곱한 결과에 편향(Bias)을 더한 것을 출력값으로 갖습니다.

torch.nn.Linear는 선형 계층을 정의할 때 사용합니다. 이 계층은 입력과 출력의 차원 수를 인자로 지정해줍니다. 예를 들어, 입력이 3개의 차원을 갖고 출력이 2개의 차원을 갖는 선형 계층을 정의하려면 `torch.nn.Linear(3,2)과 같이 정의할 수 있습니다. 선형 계층은 입력값과 가중치를 곱한 결과에 편향을 더해 출력값을 생성합니다.

# torch.nn.MSELoss
`torch.nn.MSELoss`는 예측값과 정답값의 차이를 제곱(Mean Squard Error, MSE)을 손실 함수로 정의합니다. 이 손실 함수는 신경망을 이용해 `회귀(Regression)`문제를 풀 때 자주 사용됩니다.

`torch.nn.MSELoss`는 신경망의 출력과 정답값의 차이를 측정하는 함수로, 이 차이는 예측값과 정답값의 차이의 제곱을 의미합니다. 예를 들어, 신경망이 주어진 입력으로부터 출력값이 [1,2,3]이고 정답값이 [0,1,2]인 경우 손실 함수의 값은  (1-0)^2 + (2-1)^2 + (3-2)^2 = 2가 됩니다.

torch.nn.MSELoss의 사용법은 다음과 같습니다.
```py
import torch
import torch.nn as nn

# 손실 함수 객체 생성
loss_fn = nn.MSELoss()

# 예측값과 정답값
predictions = torch.tensor([1, 2, 3])
target = torch.tensor([0, 1, 2])

# 손실 함수의 값 계산
loss = loss_fn(predictions, target)
print(loss)  # tensor(2.)
```

# parameters()
파이토치의 `parameters()`는 신경망 모듈(nn.Module)의 학습 가능한 매개변수들을 읽어오는 역할을 합니다.

신경망 모델은 입력값과 출력값을 연결하는 가중치와 편향 등의 학습 가능한 매개변수들을 갖고 있습니다. 이러한 학습 가능한 매개변수들은 신경망의 성능을 결정하는 중요한 요소이며, 신경망을 학습할 때는 이들 매개변수들을 조정하여 손실 함수의 값을 최소화하는 것을 목표로 합니다.

# torch.nn.Conv2d
`torch.nn.Conv2d`는 Pytorch에서 2D 합성곱을 수행하는 모듈로, 이미지 처리 작업에 자주 사용됩니다. `torch.nn` 모듈의 일부로, Pytorch에서 신경망을 작업하는데 필요한 다양한 고수준 구성 요소를 제공합니다.

`Conv2d` 모듈은 2D 합성곱을 입력에 적용하며, 이미지에서 공간 계층 특징을 학습하는데 사용됩니다. 이 모듈은 입력 채널 수, 출력 채널 수, 커널 크기, stride, padding 등의 다양한 매개변수를 사용합니다.

예를 들어, 아래 코드는 출력 채널 수가 64, 커널 크기가 3x3, stride가 1x1, 그리고 padding 1x1인 2D 합성곱 계층을 생성합니다.

```py
conv = torch.nn.Conv2d(in_channels=3, out_channels=64, kernel_size=3, stride=1, padding=1)
```
이 계층의 입력은 (batch_size, in_channels, height, width) 차원을 가진 4D 텐서가 되고, 출력은 (batch_size, out_channels, new_height, new_width) 차원을 가진 4D 텐서가 됩니다.

# torch.nn.ReLU
`torch.nn.ReLU`는 Pytorch의 ReLU함수를 적용하는 모듈입니다.
ReLU 함수는 입력이 0보다 작을 경우 0, 0보다 클 경우 입력 그대로를 출력합니다. 이 함수는 일반적으로 신경망에서 활성화 함수로 사용됩니다. ReLU는 매우 간단하고 연산이 빠르며, 신경망의 학습 속도를 향상시키는 등의 이점이 있습니다.

다음 코드는 ReLU 함수를 적용하는 간단한 예제입니다.
```py
import torch.nn as nn
relu = nn.ReLU()
input = torch.randn(3, 5)
output = relu(input)
```

이 코드는 3x5 크기의 무작위 입력 텐서를 생성한 후, ReLU 함수를 적용하여 출력 텐서를 얻습니다. 입력 텐서의 각 원소는 0보다 작으면 0으로 대체되며, 0보다 크면 그대로 유지됩니다.
